src/wallet/wallet_tx_builder.cpp

```diff
@@ -289,59 +509,92 @@ InputSelectionResult WalletTxBuilder::ResolveInputsAndPayments(
 
-    // TODO: the set of recipient pools is not quite sufficient information here; we should
-    // probably perform note selection at the same time as we're performing resolved payment
-    // construction above.
-    if (!spendableMut.LimitToAmount(targetAmount, dustThreshold, resolved.GetRecipientPools())) {
+    // Determine the target totals
+    CAmount sendAmount{0};
+    for (const auto& payment : payments) {
+        sendAmount += payment.GetAmount();
+    }
+
+    CAmount finalFee;
+    CAmount targetAmount;
+    std::optional<ChangeAddress> changeAddr;
+    if (fee.has_value()) {
+        finalFee = fee.value();
+        targetAmount = sendAmount + finalFee;
+        // TODO: the set of recipient pools is not quite sufficient information here; we should
+        // probably perform note selection at the same time as we're performing resolved payment
+        // construction above.
+        bool foundSufficientFunds = spendableMut.LimitToAmount(
+                targetAmount,
+                dustThreshold,
+                resolved.GetRecipientPools());
         CAmount changeAmount{spendableMut.Total() - targetAmount};
-        return InvalidFundsError(
-                spendableMut.Total(),
-                changeAmount > 0 && changeAmount < dustThreshold
-                // TODO: we should provide the option for the caller to explicitly
-                // forego change (definitionally an amount below the dust amount)
-                // and send the extra to the recipient or the miner fee to avoid
-                // creating dust change, rather than prohibit them from sending
-                // entirely in this circumstance.
-                // (Daira disagrees, as this could leak information to the recipient
-                // or publicly in the fee.)
-                ? InvalidFundsReason(DustThresholdError(dustThreshold, changeAmount))
-                : InvalidFundsReason(InsufficientFundsError(targetAmount)));
+        if (!foundSufficientFunds) {
+            return tl::make_unexpected(
+                    ReportInvalidFunds(
+                            spendableMut,
+                            false,
+                            finalFee,
+                            dustThreshold,
+                            targetAmount,
+                            changeAmount));
+        }
+        if (changeAmount > 0) {
+            changeAddr = GetChangeAddress(
+                    wallet,
+                    selector,
+                    spendableMut,
+                    resolved,
+                    strategy,
+                    afterNU5);
+            AddChangePayment(resolved, changeAddr.value(), changeAmount);
+        }
+    } else {
+        auto limitResult = IterateLimit(wallet, selector, strategy, sendAmount, dustThreshold, spendableMut, resolved, afterNU5);
+        if (limitResult.has_value()) {
+            std::tie(spendableMut, finalFee, changeAddr) = limitResult.value();
```

Comment:
I had originally suggested
```rust
    auto& [spendableMut, finalFee, changeAddr] = limitResult.value();
```
The difference is that this declares fresh local variables. The suggestion doesn't apply in this case because `spendableMut`, `finalFee`, and `changeAddr` are declared above this scope in a way that can't trivially be refactored. It technically could be refactored, but it isn't worth it here.

(On the other hand, this method's code is way too long, and splitting it up might present opportunities to simplify, for example I would consider moving the fee and change address selection into a method on `SpendableInputs` possibly. But not needed for this PR.)